### ISSUE TITLE
Using non-QWERTY keyboards with PBA

### ISSUE THREAD

**Issue by tirililtova** [Drive-Trust-Alliance/sedutil#161]
Hi,

The PBA only recognize QWERTY key map (when typing password at boot). Is there any chance that we could have international versions (I'm using French AZERTY keyboard) one day? Or is there any way to configure the PBA for that?

In all case, I suggest this should appear as a visible warning in the documentation, it can be very confusing and scary when entering the password if you're not aware of this (I defined a password in Windows 10 using an AZERTY keyboard and it was fine, but I have to use the QWERTY key map when I unlock the system at pre-boot...)

Thanks,
Tirilil

**Comment by cristim** [comment]
I think @OliverO2 got this working with the German keyboard layout, shouldn't be much difference.

**Comment by r0m30** [comment]
Yes, this is a problem, and I'm not sure what the solution is, I need to research how the live CD creators handle this.  Where do you think the warning should go where it will have the best chance of being seen?

For now, If you want to build a custom PBA with support for your keyboard you can configure buildroot to set the default keyboard to something other than us english.

**Comment by cristim** [comment]
We could have custom PBA images for each keyboard layout, or an easy way to modify the main PBA image with any layout right before writing it to the SSD.

A TinyCore PBA image may be easier to remaster than the current buildroot one.

**Comment by r0m30** [comment]
Yes, custom PBAs are an option, I'd like to figure out a way to support non-US keyboards without excessively delaying the boot process at some point.

**Comment by tirililtova** [comment]
If think the warning about non-QWERTY keyboards should appear in the main page that explains how to set the password protection on your system:

https://github.com/Drive-Trust-Alliance/sedutil/wiki/Encrypting-your-drive

This is where the PBA download and installation is described, so the keyboard warning seems to be relevant here, probably at the "Download the PBA..." section. Hopefully when non-QWERTY keyboard custom versions of the PBA are available (I think that would be great!), one should have the choice to download the version adapted to the used keyboard.

By the way, if I remember right, I noticed the QWERTY key map when testing the rescue system:

https://github.com/Drive-Trust-Alliance/sedutil/wiki/Test-the-Rescue-system

so ideally this rescue system should be available for various keyboard layouts too :-)

**Comment by r0m30** [comment]
I put a notice on the Encrypting page and the command syntax page.

Yes, both the PBA and the rescue system use the us_english keyboard.

**Comment by gomoku** [comment]
https://github.com/Drive-Trust-Alliance/sedutil/issues/114

**Comment by Zulgrib** [comment]
@OliverO2 Would this work as is if replacing -k qwertz/de-latin1 by azerty/fr-latin9 for example ?

**Comment by OliverO2** [comment]
That might work if Tiny Core Linux includes a `azerty/fr-latin9` keymap. Unfortunately, it seems that Tine Core does not provide exhaustive documentation on that topic.

Actually, I am no longer using Tiny Core since I have integrated PBA generation with local keyboard support (and more) into the current master branch of [Relax\-and\-Recover (ReaR)](https://github.com/rear/rear).

If you are a Linux user, ReaR will generate the PBA from the Linux installation you already use (including support for your local keyboard layout and more). So there is no need to download an extra Linux distribution or compile a kernel.

If you want to try it out,

- make sure that [DTA's `sedutil-cli` (version 1.15.1)](https://github.com/Drive-Trust-Alliance/exec/blob/master/sedutil_LINUX.tgz?raw=true) is available in root’s search path,
- clone or download the [current master branch of rear/rear](https://github.com/rear/rear),
- change your working directory to the rear base directory,
- create a file `etc/rear/site.conf` with the following line:
```
OUTPUT=RAWDISK
```
- run `sudo usr/sbin/rear -v mkopalpba`,
- find the PBA image in `var/lib/rear/TCG-Opal-PBA/$HOSTNAME/TCG-Opal-PBA-$HOSTNAME.raw`,
- `dd` the PBA image onto your USB stick for testing.

[Chapter 13 of the User Guide](https://github.com/rear/rear/blob/master/doc/user-guide/13-tcg-opal-support.adoc) explains how to use the PBA and administer SEDs.
